MariaDB/server
# [MDEV-30854] Do not use " as string delimiter in mariadb-tzinfo-to-sql

- [x] *The Jira issue number for this PR is: MDEV-30854*


## Description
If SQL_MODE contains ANSI_QUOTES (https://mariadb.com/kb/en/sql-mode/), then
the double-quote character (") is not a legal string delimiter.

In https://github.com/MariaDB/server/commit/13e77930e615f05cc74d408110e887b00e1abcc9#diff-a333d4ebb2d73b6361ef7dfebc86d883f7e19853b4a9eb85984b039058fae47cR2431-R2435,
@grooverdan introduced a case where the double-quote character would be used as
a string delimiter in the SQL queries generated by mariadb-tzinfo-to-sql.

This tool tool generates SQL queries which should be able to run on any
MariaDB server of the matching version.  Therefore, it should be extremely
conservative in the SQL that it outputs, in order to maximize the chance
that it can run regardless of the build or execution environment of the
server.

All new code of the whole pull request, including one or several files that are
either new files or modified ones, are contributed under the BSD-new license. I
am contributing on behalf of my employer Amazon Web Services, Inc.

## How can this PR be tested?

The preexisting MTR test `mysql_tzinfo_to_sql_symlink.test` 
actually [runs `mariadb-tzinfo-to-sql … > tz.sql` and executes the SQL which it outputs](https://github.com/MariaDB/server/blob/10.6/mysql-test/main/mysql_tzinfo_to_sql_symlink.test#L32-L35).
It continues to pass with minor modifications reflecting the expected changes in the SQL syntax; this
demonstrates that the generated SQL continues to work (in the environment where it is tested) after this PR.

See MDEV-18778, MDEV-28263, and MDEV-28782 for previous cases where MariaDB
has FAILED TO ENSURE that the generated timezone.sql actually works in
different build and execution environments. More test coverage is clearly
needed here to proactively detect such cases.




## Basing the PR against the correct MariaDB version


## Backward compatibility

This PR _improves_ compatibility, backwards and otherwise, because it makes the `mariadb-tzinfo-to-sql` output
compatible with MariaDB's parsing regardless of whether or not `SQL_MODE` includes `ANSI_QUOTES`.

## PR quality check
